File: packages/physiology-engine/src/speechGateGolden.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";

import { SpeechGate } from "./speechGate.js";

interface Case {
  readonly name: string;
  readonly note: string;
  readonly levels: readonly number[];
  readonly transmit: readonly boolean[];
}

const vectors = JSON.parse(
  readFileSync(
    new URL("../../../tests/golden/speech-gate/vectors.json", import.meta.url),
    "utf8",
  ),
) as { readonly cases: readonly Case[] };

describe("speech gate golden vectors", () => {
  // The same vectors are consumed by the Swift and Kotlin ports. Three
  // independent implementations agreeing on committed data is what parity
  // means here; agreeing on prose is not.
  it.each(vectors.cases.map((entry) => [entry.name, entry] as const))(
    "%s",
    (_name, entry) => {
      const gate = new SpeechGate();
      const actual = entry.levels.map((level) => gate.accept(level).transmit);
      expect(actual, entry.note).toEqual([...entry.transmit]);
    },
  );
});

File: apps/android/app/src/test/java/com/rafaypair/android/data/network/SpeechGateGoldenTest.kt
package com.rafaypair.android.data.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The same committed vectors the TypeScript and Swift ports consume.
 *
 * Three independent implementations agreeing on data is what parity means here;
 * agreeing on prose is not.
 */
class SpeechGateGoldenTest {
    @Serializable
    private data class GateCase(
        val name: String,
        val note: String,
        val levels: List<Double>,
        val transmit: List<Boolean>,
    )

    @Serializable
    private data class GateVectors(val cases: List<GateCase>)

    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun `matches the golden vectors`() {
        val path = "speech-gate/vectors.json"
        val stream = requireNotNull(javaClass.classLoader?.getResourceAsStream(path)) {
            "Missing golden vector $path on the test classpath"
        }
        val vectors = stream.use { json.decodeFromString<GateVectors>(it.reader().readText()) }
        assertTrue(vectors.cases.isNotEmpty())

        vectors.cases.forEach { entry ->
            val gate = SpeechGate()
            val actual = entry.levels.map { gate.accept(it).transmit }
            assertEquals("${entry.name}: ${entry.note}", entry.transmit, actual)
        }
    }
}
